src/Lib.P2P/Transports/Tcp.cs

```diff
@@ -0,0 +1,293 @@
+#region LICENSE
+
+/**
+* Copyright (c) 2019 Catalyst Network
+*
+* This file is part of Catalyst.Node <https://github.com/catalyst-network/Catalyst.Node>
+*
+* Catalyst.Node is free software: you can redistribute it and/or modify
+* it under the terms of the GNU General Public License as published by
+* the Free Software Foundation, either version 2 of the License, or
+* (at your option) any later version.
+*
+* Catalyst.Node is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Catalyst.Node. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#endregion
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Sockets;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Common.Logging;
+using MultiFormats;
+
+namespace Lib.P2P.Transports
+{
+    /// <summary>
+    ///   Establishes a duplex stream between two peers
+    ///   over TCP.
+    /// </summary>
+    /// <remarks>
+    ///   <see cref="ConnectAsync"/> determines the network latency and sets the timeout
+    ///   to 3 times the latency or <see cref="MinReadTimeout"/>.
+    /// </remarks>
+    public sealed class Tcp : IPeerTransport
+    {
+        private static readonly ILog Log = LogManager.GetLogger(typeof(Tcp));
+
+        /// <summary>
+        ///  The minimum read timeout.
+        /// </summary>
+        /// <value>
+        ///   Defaults to 3 seconds.
+        /// </value>
+        private static readonly TimeSpan MinReadTimeout = TimeSpan.FromSeconds(3);
+
+        /// <inheritdoc />
+        public async Task<Stream> ConnectAsync(MultiAddress address,
+            CancellationToken cancel = default)
+        {
+            var port = address.Protocols
+               .Where(p => p.Name == "tcp")
+               .Select(p => int.Parse(p.Value))
+               .First();
+            
+            var ip = address.Protocols
+               .FirstOrDefault(p => p.Name == "ip4" || p.Name == "ip6");
+            
+            if (ip == null)
+            {
+                throw new ArgumentException($"Missing IP address in '{address}'.", nameof(address));
+            }
+            
+            var socket = new Socket(
+                ip.Name == "ip4" ? AddressFamily.InterNetwork : AddressFamily.InterNetworkV6,
+                SocketType.Stream,
+                ProtocolType.Tcp);
+
+            var latency = MinReadTimeout; // keep compiler happy
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove this useless assignment to local variable 'latency'.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)